src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5539,12 +5539,24 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
   Label DONE, SAME;
   Register tmp1 = rscratch1;
   Register tmp2 = rscratch2;
-  Register cnt2 = tmp2;  // cnt2 only used in array length compare
   int elem_per_word = wordSize/elem_size;
   int log_elem_size = exact_log2(elem_size);
+  int klass_offset  = arrayOopDesc::klass_offset_in_bytes();
   int length_offset = arrayOopDesc::length_offset_in_bytes();
   int base_offset
     = arrayOopDesc::base_offset_in_bytes(elem_size == 2 ? T_CHAR : T_BYTE);
+  // When the base is not aligned to 8 bytes, then we let
+  // the compare loop include the array length, and skip
+  // the explicit comparison of length.
+  bool length_is_8aligned = is_aligned(length_offset, BytesPerWord);
+  assert(is_aligned(base_offset, BytesPerWord) || is_aligned(length_offset, BytesPerWord),
+         "base_offset or length_offset must be 8-byte aligned");
+  assert(is_aligned(base_offset, BytesPerWord) || base_offset == length_offset + BytesPerInt,
+         "base_offset must be 8-byte aligned or no padding between base and length");
+  int start_offset = length_is_8aligned ? length_offset : klass_offset;
+  assert(is_aligned(start_offset, BytesPerWord), "start offset must be 8-byte-aligned");
+  int extra_length = length_is_8aligned ? base_offset - length_offset : base_offset - klass_offset;
```

Comment:
Some of these asserts and comments seem a little bit out of place, inaccurate and/or incomplete with the new change to include the klass in the comparison. 

```suggestion
  // When the length offset is not aligned to 8 bytes, 
  // then we let the compare loop include the klass
  // and length, otherwise start from the length.
  bool length_is_8aligned = is_aligned(length_offset, BytesPerWord);
  int start_offset = length_is_8aligned ? length_offset : klass_offset;
  assert(is_aligned(start_offset, BytesPerWord), "start offset must be 8-byte-aligned");
  int extra_length =  base_offset - start_offset;
  assert(!length_is_8aligned || extra_length == BytesPerInt,
         "no padding between length and base");
  assert(length_is_8aligned || extra_length == BytesPerWord,
         "no padding between klass, length and base");
```

Reply:
Also right now the `length_is_8aligned` very much now just look like a `!oopDesc::has_klass_gap()` check.

The validity of choosing to start from the klass when `!length_is_8aligned` is that we must then be using `UseCompressedClassPointers /* && !UseCompactObjectHeaders  for Lilliput */`. Until Lilliput starting from the Klass is valid for both `+UseCompressedClassPointers` and `-UseCompressedClassPointers` as the klass, length and base will always be tightly packed (no padding) for byte and char type arrays. 

In all modes what we effectively do is `int start_offset = align_down(length_offset, BytesPerWord )`. Not sure if the intent gets clearer then. Something along the lines of:
```suggestion
  // When the length offset is not aligned to 8 bytes,
  // then we align it down, this is valid as the new
  // offset will always be the klass which is the same
  // for type arrays.
  int start_offset = align_down(length_offset, BytesPerWord);
  int extra_length =  base_offset - start_offset;
  assert(start_offset == length_offset || start_offset == klass_offset, 
         "start offset must be 8-byte-aligned or be the klass offset");
  assert(base_offset != start_offset, "must include the length field");
```

Reply:
Right, not assuming layout of Klass* seems saner.
BTW, including the Klass* is not only safe because it's only char[] or byte[], it's also already guaranteed to be the same by the calling APIs. And even if it were not (arraycopy-style) - if the Klass* could be different, then array-equality would have to check this, also, anyway.

Reply:
Maybe we are talking about different things, I am also unsure which method is actually intrinsified here. If it actually is `java.util.Arrays.equals(...)` then it does have the property that it does not really care about the klass. (But the method overloading ensures that you only compare the same type array object with another of the object of the same type).

```Java
class Main {
  public static void main(String args[]) {
    final int size = 10;
    Integer iArray[] = new Integer[size];
    Object oArray[] = new Object[size];

    System.out.println(iArray.equals(oArray));                   // Prints "false"
    System.out.println(oArray.equals(iArray));                   // Prints "false"
    System.out.println(java.util.Arrays.equals(iArray, oArray)); // Prints "true"
    System.out.println(iArray.getClass() == oArray.getClass());  // Prints "false"
  }
};
```